### PR TITLE
Fix record symbol finding & expression evaluation

### PR DESCRIPTION
**Pull request by mavam**
This PR contains a couple of refactoring changes centered around expression evaluation and record symbol finding.

### CODE REVIEW

**Comment by mavam** [comment]
FYI: Jenkins won't be happy with this PR until #229 is merged.

**Comment by mavam** [comment]
I think Jenkins is lying. All pass tests on my macOS dev box.

**Comment by Neverlord** [comment]
Why should it?

```
+----------------------------------------------------------------------+

                              table_index

+----------------------------------------------------------------------+



- integer values

  -> inhaling unit test suite events [line 100]

  -> building slices of 100 events each [line 121]

  -> generate table layout for flat integer type [line 67]

  -> ingest test data (integers) [line 71]

  -> verify table index [line 81]

  -> (automatically) persist table index and restore from disk [line 93]

  -> verify table index again [line 95]

  -> 35 checks took 582 ms



- record type

  -> generate table layout for record type [line 100]

  -> ingest test data (records) [line 107]

  -> verify table index [line 124]

  -> (automatically) persist table index and restore from disk [line 131]

  -> verify table index again [line 133]

  -> 8 checks took 1 ms



- bro conn logs

  -> generate table layout for bro conn logs [line 138]

  -> ingest test data (bro conn log) [line 141]

  -> verify table index [line 144]

  -> (automatically) persist table index and restore from disk [line 161]

  -> verify table index again [line 163]

  -> 24 checks took 427 ms

 (14/10)



+----------------------------------------------------------------------+

                                summary

+----------------------------------------------------------------------+



                        suites:  1

                        tests:   3

                        checks:  67 (57/10)

                        time:    1.1 s

                        success: 85%



                        failed tests:

                        - table_index:bro conn logs



+----------------------------------------------------------------------+
```

The unit test fail on all setups, so it can't be just some issues with a single box.

**Comment by mavam** [comment]
I think this must somehow be stale output. I cannot reproduce the failing test.

I've just pulled the latest CAF master again and the tests still pass locally. Since the Jenkins macOS box shouldn't be different my local setup, my hunch is that there is some sort of disconnect.

**Comment by Neverlord** [comment]
@mavam the tests fail on my machine as well:

```
$ ./build/bin/vast-test -r0 -v4 -s "^indexer$"
+----------------------------------------------------------------------+
                                indexer
+----------------------------------------------------------------------+

- integer rows
  -> inhaling unit test suite events [line 100]
** ../libvast/test/fixtures/events.hpp:93   !e
** ../libvast/test/fixtures/events.hpp:94   e.error() == ec::end_of_input
** ../libvast/test/fixtures/events.hpp:95   !events.empty()
** ../libvast/test/fixtures/events.hpp:93   !e
** ../libvast/test/fixtures/events.hpp:94   e.error() == ec::end_of_input
** ../libvast/test/fixtures/events.hpp:95   !events.empty()
** ../libvast/test/fixtures/events.hpp:93   !e
** ../libvast/test/fixtures/events.hpp:94   e.error() == ec::end_of_input
** ../libvast/test/fixtures/events.hpp:95   !events.empty()
** ../libvast/test/fixtures/events.hpp:93   !e
** ../libvast/test/fixtures/events.hpp:94   e.error() == ec::end_of_input
** ../libvast/test/fixtures/events.hpp:95   !events.empty()
** ../libvast/test/fixtures/events.hpp:93   !e
** ../libvast/test/fixtures/events.hpp:94   e.error() == ec::end_of_input
** ../libvast/test/fixtures/events.hpp:95   !events.empty()
  -> building slices of 100 events each [line 121]
** ../libvast/test/fixtures/events.cpp:194  ::caf::test::equal_to(bro_conn_log.size(), flat_log.size())
  -> ingest integer events [line 64]
  -> verify table index [line 73]
** ../libvast/test/system/indexer.cpp:75   ::caf::test::equal_to(query(":int == +1"), res(0, 3, 6))
** ../libvast/test/system/indexer.cpp:76   ::caf::test::equal_to(query(":int == +2"), res(1, 4, 7))
** ../libvast/test/system/indexer.cpp:77   ::caf::test::equal_to(query(":int == +3"), res(2, 5, 8))
** ../libvast/test/system/indexer.cpp:78   ::caf::test::equal_to(query(":int == +4"), res())
** ../libvast/test/system/indexer.cpp:79   ::caf::test::equal_to(query(":int != +1"), res(1, 2, 4, 5, 7, 8))
** ../libvast/test/system/indexer.cpp:80   ::caf::test::equal_to(query("!(:int == +1)"), res(1, 2, 4, 5, 7, 8))
** ../libvast/test/system/indexer.cpp:81   ::caf::test::equal_to(query(":int > +1 && :int < +3"), res(1, 4, 7))
  -> kill INDEXER [line 84]
  -> reload INDEXER from disk [line 87]
  -> verify table index again [line 89]
** ../libvast/test/system/indexer.cpp:75   ::caf::test::equal_to(query(":int == +1"), res(0, 3, 6))
** ../libvast/test/system/indexer.cpp:76   ::caf::test::equal_to(query(":int == +2"), res(1, 4, 7))
** ../libvast/test/system/indexer.cpp:77   ::caf::test::equal_to(query(":int == +3"), res(2, 5, 8))
** ../libvast/test/system/indexer.cpp:78   ::caf::test::equal_to(query(":int == +4"), res())
** ../libvast/test/system/indexer.cpp:79   ::caf::test::equal_to(query(":int != +1"), res(1, 2, 4, 5, 7, 8))
** ../libvast/test/system/indexer.cpp:80   ::caf::test::equal_to(query("!(:int == +1)"), res(1, 2, 4, 5, 7, 8))
** ../libvast/test/system/indexer.cpp:81   ::caf::test::equal_to(query(":int > +1 && :int < +3"), res(1, 4, 7))
  -> 30 checks took 12.13 s

- bro conn logs
  -> ingest bro conn log [line 94]
  -> verify table index [line 96]
!! ../libvast/test/system/indexer.cpp:101  ::caf::test::equal_to(rank(query("id.resp_p == 995/?")), 53u) (0 !! 53)
!! ../libvast/test/system/indexer.cpp:102  ::caf::test::equal_to(rank(query("id.resp_p == 5355/?")), 49u) (0 !! 49)
!! ../libvast/test/system/indexer.cpp:103  ::caf::test::equal_to(rank(query("id.resp_p == 995/? || id.resp_p == 5355/?")), 102u) (0 !! 102)
** ../libvast/test/system/indexer.cpp:104  ::caf::test::equal_to(rank(query("&time > 1970-01-01")), bro_conn_log.size())
** ../libvast/test/system/indexer.cpp:105  ::caf::test::equal_to(rank(query("proto == \"udp\"")), 5306u)
** ../libvast/test/system/indexer.cpp:106  ::caf::test::equal_to(rank(query("proto == \"tcp\"")), 3135u)
** ../libvast/test/system/indexer.cpp:107  ::caf::test::equal_to(rank(query("uid == \"nkCxlvNN8pi\"")), 1u)
!! ../libvast/test/system/indexer.cpp:108  ::caf::test::equal_to(rank(query("orig_bytes < 400")), 5332u) (0 !! 5332)
!! ../libvast/test/system/indexer.cpp:109  ::caf::test::equal_to(rank(query("orig_bytes < 400 && proto == \"udp\"")), 4357u) (0 !! 4357)
** ../libvast/test/system/indexer.cpp:110  ::caf::test::equal_to(rank(query(":addr == 65.55.184.16")), 2u)
** ../libvast/test/system/indexer.cpp:111  ::caf::test::equal_to(query(":addr == 169.254.225.22"), res(680, 682, 719, 720))
  -> kill INDEXER [line 114]
  -> reload INDEXER from disk [line 117]
  -> verify table index again [line 119]
!! ../libvast/test/system/indexer.cpp:101  ::caf::test::equal_to(rank(query("id.resp_p == 995/?")), 53u) (0 !! 53)
!! ../libvast/test/system/indexer.cpp:102  ::caf::test::equal_to(rank(query("id.resp_p == 5355/?")), 49u) (0 !! 49)
!! ../libvast/test/system/indexer.cpp:103  ::caf::test::equal_to(rank(query("id.resp_p == 995/? || id.resp_p == 5355/?")), 102u) (0 !! 102)
** ../libvast/test/system/indexer.cpp:104  ::caf::test::equal_to(rank(query("&time > 1970-01-01")), bro_conn_log.size())
** ../libvast/test/system/indexer.cpp:105  ::caf::test::equal_to(rank(query("proto == \"udp\"")), 5306u)
** ../libvast/test/system/indexer.cpp:106  ::caf::test::equal_to(rank(query("proto == \"tcp\"")), 3135u)
** ../libvast/test/system/indexer.cpp:107  ::caf::test::equal_to(rank(query("uid == \"nkCxlvNN8pi\"")), 1u)
!! ../libvast/test/system/indexer.cpp:108  ::caf::test::equal_to(rank(query("orig_bytes < 400")), 5332u) (0 !! 5332)
!! ../libvast/test/system/indexer.cpp:109  ::caf::test::equal_to(rank(query("orig_bytes < 400 && proto == \"udp\"")), 4357u) (0 !! 4357)
** ../libvast/test/system/indexer.cpp:110  ::caf::test::equal_to(rank(query(":addr == 65.55.184.16")), 2u)
** ../libvast/test/system/indexer.cpp:111  ::caf::test::equal_to(query(":addr == 169.254.225.22"), res(680, 682, 719, 720))
  -> 22 checks took 8.96 s
 (12/10)

+----------------------------------------------------------------------+
                                summary
+----------------------------------------------------------------------+

                        suites:  1
                        tests:   2
                        checks:  52 (42/10)
                        time:    21.9 s
                        success: 80%

                        failed tests:
                        - indexer:bro conn logs

+----------------------------------------------------------------------+
```

```
$ git rev-parse HEAD
8e21425c60e8028c68d7d2b5520f403193acb2f4
$ (cd ../caf && git rev-parse HEAD)
696736e4ad23765d2920d47488acd9703d3494c0
```

**Comment by Neverlord** [comment]
For the sake of completeness, here's my local configuration for the build:

```
$ cat build/config.status 
#!/bin/sh

# Switch to the source of this build directory.
cd "/Users/neverlord/workspace/vast"

# Invoke the command to configure this build.
./configure "--generator=Ninja" "--build-type=debug" "--enable-asan" "--with-caf=/Users/neverlord/workspace/caf/build"
```

```
$ cat ../caf/build/config.status 
#!/bin/sh

# Switch to the source of this build directory.
cd "/Users/neverlord/workspace/caf"

# Invoke the command to configure this build.
./configure --generator=Ninja --build-type=debug --with-runtime-checks --enable-asan --with-log-level=trace --no-opencl
```

**Comment by mavam** [comment]
Thanks for sharing your build setup. For some reason, the bug didn't show up in my `RelWithDebInfo` build. But I could reproduce and fix it in plain `Debug` build.

Anyway, this PR is ready to be reviewed.

File: libvast/src/type.cpp [hunk 1]
```diff
@@ -207,155 +209,146 @@ const record_type::each::range_state& record_type::each::get() const {
...
   offset off_;
 };
 
 } // namespace <anonymous>
 
-std::vector<std::pair<offset, key>> record_type::find(const key& k) const {
-  return finder{k, finder::exact, name()}(*this);
+std::vector<std::pair<offset, std::string>>
+record_type::find(std::string_view key) const {
+    return finder<mode::any>{key}(*this);
```
Comment (Neverlord) [inline libvast/src/type.cpp]: Indentation (4 instead of 2 spaces).

File: libvast/src/type.cpp [hunk 2]
```diff
@@ -207,155 +209,146 @@ const record_type::each::range_state& record_type::each::get() const {
...
 }
 
-std::vector<std::pair<offset, key>>
-record_type::find_prefix(const key& k) const {
-  return finder{k, finder::prefix, name()}(*this);
+std::vector<std::pair<offset, std::string>>
+record_type::find_prefix(std::string_view key) const {
+    return finder<mode::prefix>{key}(*this);
```
Comment (Neverlord) [inline libvast/src/type.cpp]: Indentation (4 instead of 2 spaces).

File: libvast/src/type.cpp [hunk 3]
```diff
@@ -207,155 +209,146 @@ const record_type::each::range_state& record_type::each::get() const {
...
 }
 
-std::vector<std::pair<offset, key>>
-record_type::find_suffix(const key& k) const {
-  return finder{k, finder::suffix, name()}(*this);
+std::vector<std::pair<offset, std::string>>
+record_type::find_suffix(std::string_view key) const {
+    return finder<mode::suffix>{key}(*this);
```
Comment (Neverlord) [inline libvast/src/type.cpp]: Indentation (4 instead of 2 spaces).

File: libvast/test/expression.cpp [hunk 4]
```diff
@@ -36,7 +36,7 @@ using namespace vast;
 struct fixture {
   fixture() {
     // expr0 := !(x.y.z <= 42 && &foo == T)
-    auto p0 = predicate{key_extractor{{"x", "y", "z"}}, less_equal, data{42}};
+    auto p0 = predicate{key_extractor{"x.y.z"}, less_equal, data{42}};
```
Comment (Neverlord) [inline libvast/test/expression.cpp]: Reads much nicer now. 😉

File: libvast/test/pattern.cpp [hunk 5]
```diff
@@ -22,6 +22,7 @@
 
 using namespace vast;
 using namespace std::string_literals;
+using namespace std::string_view_literals;
```
Comment (Neverlord) [inline libvast/test/pattern.cpp]: I don't spot any `""sv`, is this really used?

File: libvast/src/expression.cpp [hunk 6]
```diff
@@ -33,7 +33,7 @@ bool operator<(const attribute_extractor& x, const attribute_extractor& y) {
 
 // -- key_extractor ------------------------------------------------------------
 
-key_extractor::key_extractor(vast::key k) : key{std::move(k)} {
+key_extractor::key_extractor(std::string k) : key{std::move(k)} {
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/expression.cpp]: I don't think so. Expressions get passed around and serialized. A `key_extractor` is a predicate operand and therefore needs to have owning semantics.

File: libvast/src/pattern.cpp [hunk 7]
```diff
@@ -20,27 +20,95 @@
...
+bool pattern::search(std::string_view str) const {
   return std::regex_search(str.begin(), str.end(), std::regex{str_});
 }
 
 const std::string& pattern::string() const {
   return str_;
 }
 
+pattern& pattern::operator+=(const pattern& other) {
+  str_ += other.str_;
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/pattern.cpp]: ✅

File: libvast/src/table_index.cpp [hunk 8]
```diff
@@ -21,7 +21,18 @@
 #include "vast/save.hpp"
 #include "vast/table_slice.hpp"
 
+#include "vast/detail/overload.hpp"
+#include "vast/detail/string.hpp"
+
 namespace vast {
+namespace {
+
+// Translates a key to directory.
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/table_index.cpp]: ✅

File: libvast/src/type.cpp [hunk 9]
```diff
@@ -159,11 +160,12 @@ record_type::record_type(std::initializer_list<record_field> xs)
   // nop
 }
 
-key record_type::each::range_state::key() const {
-  vast::key k(trace.size());
-  for (size_t i = 0; i < trace.size(); ++i)
-    k[i] = trace[i]->name;
-  return k;
+std::string record_type::each::range_state::key() const {
+  std::vector<std::string> result;
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/type.cpp]: Indeed, so much nicer.

File: libvast/src/type.cpp (hunk 1 above)
Comment (mavam) [reply libvast/src/type.cpp]: ✅

File: libvast/src/type.cpp (hunk 2 above)
Comment (mavam) [reply libvast/src/type.cpp]: ✅

File: libvast/src/type.cpp (hunk 3 above)
Comment (mavam) [reply libvast/src/type.cpp]: ✅

File: libvast/test/pattern.cpp (hunk 5 above)
Comment (mavam) [reply libvast/test/pattern.cpp]: ✅

File: libvast/src/type.cpp [hunk 10]
```diff
@@ -207,155 +209,146 @@ const record_type::each::range_state& record_type::each::get() const {
...
-        result.push_back(std::move(p));
-      trace_.pop_back();
+      auto prev_trace_size = trace_.size();
+      trace_ += trace_.empty() ? f.name : '.' + f.name;
+      auto sub_result = visit(*this, f.type);
+      result.insert(result.end(),
+                    std::make_move_iterator(sub_result.begin()),
+                    std::make_move_iterator(sub_result.end()));
+      trace_.resize(prev_trace_size);
       ++off_.back();
```
[the remark this answers is not in the document]
Comment (mavam) [reply libvast/src/type.cpp]: Yeah, I agree that this is confusing. We have one other occurrence of this kind in this file:
```cpp
while (++state_.offset.back() == records_.back()->fields.size()) {
```
How would you rewrite this one? I can't come up with anything short and sweet that doesn't uglify it. This for example,
```cpp
for (state_.offset.back() +=1;
     state_.offset.back() == records_.back()->fields.size();
     state_.offset.back() +=1) {
  // ...
```
is just too much noise for my taste.

**Pull request merged**